dataset.py

```diff
@@ -0,0 +1,189 @@
+                'Unknown source/target type. Should be one of '
+                f'"speech" or "text", got: {source_target_config}.'
+            )
+    return matches
+
+def _matching_pairs(row,
+                    config,
+                    source_preprocess_fn,
+                    target_preprocess_fn):
+    """Find all source/target pairs that match the configuration."""
+    source_items = _matching_items(row, config['source'])
+    target_items = _matching_items(row, config['target'])
+    
+    for source in source_items:
+        source = source_preprocess_fn(source)
+        for target in target_items:
+            target = target_preprocess_fn(target)
+            yield {
+                'source': source['value'],
+                'source_language': source['language'],
+                'target': target['value'],
+                'target_language': target['language'],
+            }
+
+def _create_generator(config):
+    source_preprocess_fn = lambda x: x
+    target_preprocess_fn = lambda x: x
+    huggingface_datasets = _load_huggingface_datasets(config)
+    for ds in huggingface_datasets:
+        for row in ds:
+            for match in _matching_pairs(row, config,
+                                         source_preprocess_fn,
+                                         target_preprocess_fn):
+                yield {k: match[k] for k in ['source', 'target']}
+
+def create(config):
+    """
+    Create a dataset from the given configuration.
+
+    Args:
+      huggingface_load : Dict containing keyword arguments to HuggingFace
+          datasets.load_dataset(), or a list of dicts to load multiple
+          datasets.
+      source: Dict containing source specification, as below.
+      target: Dict containing target specification, as below.
+      shuffle: Whether to shuffle the data after loading (default False).
+
+    Source and target configuration:
+      language: Either an ISO 639-2 language code (e.g. 'eng', 'lug'),
+          or a list of codes.
+      type: 'text' or 'speech'.
+      recording_type: In the case of audio, 'studio', 'natural' or
+          'any' (default).
+      preprocessing: list of any functions that should be applied at
+          load time (done once, same output every epoch).
+      preprocessing_on_the_fly: list of any functions that should be applied
+          subsequently, on the fly (e.g. augmentation, for different output
+          every epoch).
+
+    Returns:
+      dataset: A datasets.Dataset object with attributes `source` and `target`.
+    """
+    # TODO: checks on configuration to make sure it's valid.
+    # TODO: make sample rate configurable.
+    
+    for language in [config[s]['language'] for s in ['source', 'target']]:
+        if ',' in language and isinstance(language, str):
+            raise ValueError(
+                'A list of languages has been specified in config as a string: '
+                f'{language}. Change to [{language}] to make it a list.')
```

Comment:
are we checking for the comma to ensure the languages are always passed as a list? we should maybe make this more clear

Reply:
Good point, this should be more clear. I added a comment explaining what the purpose is here. Basically if someone specifies `language: lug, nyn, ach` (without square brackets), then that actually evaluates to the string "lug, nyn, ach" and not a list.